arraycontext/context.py

```diff
@@ -123,10 +113,64 @@
 from pytools.tag import Tag
 
 
-DeviceArray = Any
-DeviceScalar = Any
+# {{{ typing
+
 _ScalarLike = Union[int, float, complex, np.generic]
 
+try:
+    from typing import Protocol
+except ImportError:
+    from typing_extensions import Protocol                  # type: ignore[misc]
+
+
+class Array(Protocol):
+    """A :class:`~typing.Protocol` for the array type supported by
+    :class:`ArrayContext`.
+
+    This is meant to aid in typing annotations. For a explicit list of
+    supported types see :attr:`ArrayContext.array_types`.
+
+    .. attribute:: shape
+    .. attribute:: dtype
+    """
+
+    @property
+    def shape(self) -> Tuple[int, ...]:
+        ...
+
+    @property
+    def dtype(self) -> "np.dtype[Any]":
+        ...
+
+
+class Scalar(Protocol):
+    """A :class:`~typing.Protocol` for the scalar type supported by
+    :class:`ArrayContext`.
+
+    In :mod:`numpy` terminology, this is just an array with a shape of ``()``.
+
+    This is meant to aid in typing annotations. For a explicit list of
+    supported types see :attr:`ArrayContext.array_types`.
+
+    .. attribute:: shape
+    .. attribute:: dtype
+    """
+
+    @property
+    def shape(self) -> Tuple[()]:
+        ...
+
+    @property
+    def dtype(self) -> "np.dtype[Any]":
+        ...
+
+
+# NOTE: backwards compatibility
+DeviceArray = Array
+DeviceScalar = Scalar
```

Comment:
Should we do a deprecation via `__getattr__` on these?

Reply:
From a quick grep, only `DeviceScalar` is used in a few places in grudge.

Added in https://github.com/inducer/arraycontext/compare/a9252057a392bcab1d0ac79ac193e9040a60d6f9..160c66dd044932e698641e57d9f9197d239e252e.